File: pages/refactorings/replace-inheritance-with-natural-transformation/01-stack-vector/Before.hs
-- Editor history: apply edits and undos, then list the survivors
-- oldest-first. A Stack is-a list, so the whole list API is the
-- history's API: undo is list surgery like any other operation.
module Before where

import Data.List (foldl')

type Stack a = [a]

data Cmd = Edit String | Undo

build :: [Cmd] -> Stack String
build = foldl' step []
  where
    step h (Edit t) = h ++ [t]
    step h Undo     = take (length h - 1) h

replay :: [Cmd] -> [String]
replay cmds = build cmds -- compiles: a Stack *is* the list

File: pages/refactorings/replace-inheritance-with-natural-transformation/01-stack-vector/Before.scala
// Editor history: apply edits and undos, then list the survivors
// oldest-first. A Stack is-a List, so the whole List API is the
// history's API: undo is list surgery like any other operation.
object Before:
  type Stack[A] = List[A]

  enum Cmd:
    case Edit(text: String)
    case Undo

  def build(cmds: List[Cmd]): Stack[String] =
    cmds.foldLeft(Nil: Stack[String]) {
      case (h, Cmd.Edit(t)) => h :+ t
      case (h, Cmd.Undo)    => h.dropRight(1)
    }

  def replay(cmds: List[Cmd]): List[String] =
    build(cmds) // compiles only because a Stack *is* the List
